xclim/core/bootstrapping.py

```diff
@@ -0,0 +1,236 @@
+    Thus the years of the base period are affected by sampling variability.
+
+    To workaround this issue in the in-base period, we can apply the following algorithm
+    The data set of in-base period will thereafter be called ds
+    1. Split ds in an in-base of 29 years and a single year of out of base period
+    2. Add a virtual year in the constructed in-base period by repeating one year.
+    3. Get the thershold (your percentile of interest) form this 30 years in-base period
+    4. Calculate the exceedance rate of the out base period (the single year)
+    5. Repeat 28 times the steps 2, 3, 4 but each time, the repeated year of step 2 must be another year of the in-base
+    6. To obtain the final time serie of exceedance rate for the out of base year, average the 29 estimates obtained
+
+    7. Repeat the whole process for each year of the base period
+
+    TODO: show how to bootstrap a custom indice
+
+    """
+    in_base_period = da.sel(time=config.in_base_slice)
+    in_base_exceedance_rates = _bootstrap_period(
+        in_base_period, config, exceedance_function
+    )
+    if config.out_of_base_slice is None:
+        return in_base_exceedance_rates
+    out_of_base_period = da.sel(time=config.out_of_base_slice)
+    in_base_threshold = _calculate_thresholds(in_base_period, config)
+    out_of_base_exceedance = _calculate_exceedances(
+        config, exceedance_function, out_of_base_period, in_base_threshold
+    )
+    return xr.concat([in_base_exceedance_rates, out_of_base_exceedance], dim="time")
+
+
+def _bootstrap_period(
+    ds_in_base_period: DataArray,
+    config: BootstrapConfig,
+    exceedance_function: ExceedanceFunction,
+) -> DataArray:
+    period_exceedance_rates = []
+    for year_ds in ds_in_base_period.groupby("time.year"):
+        period_exceedance_rates.append(
+            _bootstrap_year(ds_in_base_period, year_ds[0], config, exceedance_function)
+        )
+    out = xr.concat(period_exceedance_rates, dim="time")
+    # workaround to ensure unit is really "days"
+    out.attrs["units"] = "d"
+    return out
+
+
+def _bootstrap_year(
+    ds_in_base_period: DataArray,
+    out_base_year: int,
+    config: BootstrapConfig,
+    exceedance_function: ExceedanceFunction,
+) -> DataArray:
+    print("out base :", out_base_year)
+    in_base = _build_virtual_in_base_period(ds_in_base_period, out_base_year)
+    out_base = ds_in_base_period.sel(time=str(out_base_year))
+    exceedance_rates = []
+    for year_ds in in_base.groupby("time.year"):
+        print(year_ds[0])
+        replicated_year = in_base.sel(time=str(year_ds[0]))
+        # it is necessary to change the time of the replicated year
+        # in order to not skip it in percentile calculation
+        replicated_year["time"] = replicated_year.time + pd.Timedelta(
+            str(out_base_year - year_ds[0]) + "y"
+        )
```

Comment:
Same comment for `year_ds[0]` as above.

Also, what if the replicated year is a leap year and this addition brings it to a non-leap year? After testing, I think everything will be offset by a day, which ain't good!

Idea : convert everything to the closest uniform calendar ( usually 'noleap', thus dropping all 29th Feb). In that case we will be able to _replace_ the time axis :

```python
replicated_year['time'] = ds_in_base_period.time.sel(time=str(out_base_year))
```

Not sure how this calendar conversion will impact the computation, Are the percentiles of day 366 really trustable anyway?

Reply:
I share your concern with the pertinence of 29th feb percentiles, I _guess_ we can skip it entirely, I'm confirming this with Christian. 
So, you suggest that we do something like ``da = calendar.convert_calendar(da, "noleap")`` at the very beginning of bootstrapping ?

Reply:
I discussed the matter with Christian P. and for icclim there is no special treatment for 29th Feb. 
The idea is that, when the **out of base year is a non-leap year** and we replicate a leap year, (thus we have an extra day in the dataset), this is compensated by when the **out of base year is a leap year** and we replicate a non leap year (thus a day is lacking).
So ideally, we should copy as is the year but make sure the time dimension is not offsetted.